File: Editorial/89_Gray_Code/TypeScript/4_Recursion_2.ts
let nextNum = 0;
function grayCode(n: number): number[] {
    let result: number[] = [];
    grayCodeHelper(result, n);
    nextNum = 0; //reset global variable after all recursions
    return result;
}
let grayCodeHelper = function (result: number[], n: number): void {
    if (n == 0) {
        result.push(nextNum);
        return;
    }
    grayCodeHelper(result, n - 1);
    // Flip the bit at (n - 1)th position from right
    nextNum = nextNum ^ (1 << (n - 1));
    grayCodeHelper(result, n - 1);
};

File: Editorial/89_Gray_Code/C/4_Recursion_2.c
int nextNum = 0;
void grayCodeHelper(int* result, int* size, int n) {
    if (n == 0) {
        result[*size] = nextNum;
        (*size)++;
        return;
    }
    grayCodeHelper(result, size, n - 1);
    nextNum = nextNum ^ (1 << (n - 1));
    grayCodeHelper(result, size, n - 1);
}
int* grayCode(int n, int* returnSize) {
    int max = 1 << n;
    int* result = (int*)malloc(sizeof(int) * max);
    *returnSize = 0;
    nextNum = 0;
    grayCodeHelper(result, returnSize, n);
    return result;
}
